crates/pop-drink/src/error.rs

```diff
@@ -0,0 +1,154 @@
+use std::fmt::Debug;
+
+pub use drink::{
+	pallet_assets::Error as AssetsError, pallet_balances::Error as BalancesError,
+	pallet_contracts::Error as ContractsError,
+};
+pub use frame_support::sp_runtime::{
+	ArithmeticError, DispatchError, ModuleError, TokenError, TransactionalError,
+};
+use scale::{Decode, Encode};
+
+fn decode<T: Decode>(data: &[u8]) -> T {
+	T::decode(&mut &data[..]).expect("Decoding failed")
+}
+
+#[derive(Encode, Decode, Debug)]
+pub enum DrinkError<E>
+where
+	E: Decode + Encode + Debug,
+{
+	Raw(DispatchError),
+	Module(E),
+}
+
+impl<E> From<DrinkError<E>> for u32
+where
+	E: Decode + Encode + Debug,
+{
+	fn from(error: DrinkError<E>) -> Self {
+		let mut value = match error {
+			DrinkError::Raw(dispatch_error) => dispatch_error.encode(),
+			DrinkError::Module(module_error) => {
+				let mut module_error = module_error.encode();
+				module_error.insert(0, 3);
+				module_error
+			},
+		};
+		value.resize(4, 0);
+		u32::from_le_bytes(value.try_into().expect("qed, resized to 4 bytes line above"))
+	}
+}
+
+impl<E> From<u32> for DrinkError<E>
+where
+	E: Decode + Encode + Debug,
+{
+	fn from(value: u32) -> Self {
+		let encoded = value.to_le_bytes();
+		match encoded {
+			[3, module_error @ ..] => DrinkError::Module(decode(&module_error)),
+			_ => DrinkError::Raw(decode(&encoded)),
+		}
+	}
+}
+
+#[track_caller]
+pub fn assert_runtime_err_inner<T, R, E>(result: Result<R, E>, expected_error: DrinkError<T>)
+where
+	T: Decode + Encode + Debug,
+	E: Into<u32>,
+{
+	let expected_code: u32 = expected_error.into();
+	let expected_error = DrinkError::<T>::from(expected_code);
+	if let Err(error) = result {
+		let error_code: u32 = error.into();
+		if error_code != expected_code {
+			panic!(
+				r#"assertion `left == right` failed
+  left: {:?}
+ right: {:?}"#,
+				DrinkError::<T>::from(error_code),
+				expected_error,
+			);
+		}
+	} else {
+		panic!(
+			r#"assertion `left == right` failed
+  left: Ok()
+ right: {:?}"#,
+			expected_error,
+		);
+	}
+}
+
+#[macro_export]
+macro_rules! assert_runtime_err {
+	($result:expr, $error:expr $(,)?) => {
+		$crate::error::assert_runtime_err_inner($result, $error);
+	};
+}
+
+#[cfg(test)]
+mod test {
+	use pop_api::primitives::Error as PopApiError;
+
+	use super::{
+		AssetsError::*,
+		BalancesError::*,
+		DrinkError::{self, *},
+	};
+
+	fn test_cases() -> Vec<(DrinkError<crate::mock::RuntimeError>, PopApiError)> {
+		use frame_support::traits::PalletInfoAccess;
+
+		use crate::mock::RuntimeError::*;
+		vec![
```

Comment:
Can we add the BadOrigin dispatcherror? Just to have a unit variant tested as well.

Reply:
Resolved https://github.com/r0gue-io/pop-drink/pull/14/commits/201a9b9eac1deb3bfcec3fea2b38d4418fb6b60a